### PR TITLE
[CALCITE-4081] Round-tripping a DECIMAL literal throws validation error

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-4081

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java [hunk 1]
```diff
@@ -1751,4 +1752,18 @@ public static RelDataType extractLastNFields(RelDataTypeFactory typeFactory,
...
+
+  public static boolean isValidDecimalValue(BigDecimal value, RelDataType toType) {
+    if (value == null) {
+      return true;
+    }
+    switch (toType.getSqlTypeName()) {
+    case DECIMAL:
+      final int intDigits = value.precision() - value.scale();
+      final int maxIntDigits = toType.getPrecision() - toType.getScale();
+      return intDigits <= maxIntDigits;
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1764]: Seems the `toType` should always be a `Decimal` type, we should add a check here. Also, can we add some doc to this method ?

File: core/src/test/resources/sql/misc.iq [hunk 2]
```diff
@@ -1689,7 +1689,7 @@ EnumerableCalc(expr#0=[{inputs}], expr#1=[123:BIGINT], EXPR$0=[$t1])
 !plan
 
 # Cast an integer literal to a decimal; note: the plan does not contain CAST
-values cast('123.45' as decimal(4, 2));
+values cast('123.45' as decimal(5, 2));
 +--------+
```
Comment (danny0405) [inline core/src/test/resources/sql/misc.iq:1693]: Why change the query ? It works in MySql and PostgresSQL.

Comment (chunweilei) [reply core/src/test/resources/sql/misc.iq:1693]: It overflows.
```
chunwei.lcw@chunwei:~$ psql
psql (12.3)
Type "help" for help.

chunwei.lcw=# select cast('111.23' as decimal(4, 2));
ERROR:  numeric field overflow
DETAIL:  A field with precision 4, scale 2 must round to an absolute value less than 10^2.
```

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java (hunk 1 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1764]: > Seems the toType should always be a Decimal type, we should add a check here

`toType` can be any type. I used `case when`. So I think it is fine.

> Also, can we add some doc to this method ?

Makes sense.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1772]: If `toType` is not decimal, how can you ensure that it has a valid precision and scale, such as `-1`.

File: core/src/test/resources/sql/misc.iq (hunk 2 above)
Comment (danny0405) [reply core/src/test/resources/sql/misc.iq:1693]: Hmm, it seems only MySql succeed, with a result of `99.99` for query `select cast('123.45' as decimal(4, 2))`

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java (hunk 1 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1772]: Do you want to throw an exception if toType is not decimal by saying "add a check here"?

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1772]: I just want to make sure the logic is robust, either a check or assert is okey from  my side.

File: core/src/test/resources/sql/misc.iq (hunk 2 above)
Comment (danny0405) [reply core/src/test/resources/sql/misc.iq:1693]: We may need to have a conclusion for this behavior change, it seems breaking.

Comment (chunweilei) [reply core/src/test/resources/sql/misc.iq:1693]: Yep.

**Pull request merged**